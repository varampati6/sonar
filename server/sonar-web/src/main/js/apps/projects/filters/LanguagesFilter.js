/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
//@flow
import React from 'react';
import { difference, sortBy } from 'lodash';
import Filter from './Filter';
import FilterHeader from './FilterHeader';
import SearchableFilterFooter from './SearchableFilterFooter';
import SearchableFilterOption from './SearchableFilterOption';
import { getLanguageByKey } from '../../../store/languages/reducer';

type Props = {
  query: {},
  languages: {},
  router: { push: ({ pathname: string, query?: {} }) => void },
  value?: Array<string>,
  facet?: {},
  isFavorite?: boolean,
  organization?: {},
  maxFacetValue?: number
};

const LIST_SIZE = 10;

export default class LanguagesFilter extends React.PureComponent {
  getSearchOptions: () => [{ label: string, value: string }];
  props: Props;
  property = 'languages';

  getSearchOptions(facet: {}, languages: {}) {
    let languageKeys = Object.keys(languages);
    if (facet) {
      languageKeys = difference(languageKeys, Object.keys(facet));
    }
    return languageKeys
      .slice(0, LIST_SIZE)
      .map(key => ({ label: languages[key].name, value: key }));
  }

  getSortedOptions(facet: {} = {}) {
    return sortBy(Object.keys(facet), [option => -facet[option], option => option]);
  }

  getFacetValueForOption = (facet: {} = {}, option: string) => facet[option];

  renderOption = (option: string) => (
    <SearchableFilterOption
      optionKey={option}
      option={getLanguageByKey(this.props.languages, option)}
    />
  );

  render() {
    return (
      <Filter
        property={this.property}
        options={this.getSortedOptions(this.props.facet)}
        query={this.props.query}
        renderOption={this.renderOption}
        value={this.props.value}
        facet={this.props.facet}
        maxFacetValue={this.props.maxFacetValue}
        isFavorite={this.props.isFavorite}
        organization={this.props.organization}
        getFacetValueForOption={this.getFacetValueForOption}
        highlightUnder={1}
        header={<FilterHeader name="Languages" />}
        footer={
          <SearchableFilterFooter
            property={this.property}
            query={this.props.query}
            options={this.getSearchOptions(this.props.facet, this.props.languages)}
            isFavorite={this.props.isFavorite}
            organization={this.props.organization}
            router={this.props.router}
          />
        }
      />
    );
  }
}
