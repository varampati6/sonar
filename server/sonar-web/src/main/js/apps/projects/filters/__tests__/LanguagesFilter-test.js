/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import LanguagesFilter from '../LanguagesFilter';

const languages = {
  java: {
    key: 'java',
    name: 'Java'
  },
  cs: {
    key: 'cs',
    name: 'C#'
  },
  js: {
    key: 'js',
    name: 'JavaScript'
  },
  flex: {
    key: 'flex',
    name: 'Flex'
  },
  php: {
    key: 'php',
    name: 'PHP'
  },
  py: {
    key: 'py',
    name: 'Python'
  }
};
const languagesFacet = { java: 39, cs: 4, js: 1 };
const fakeRouter = { push: () => {} };

it('should render the languages without the ones in the facet', () => {
  const wrapper = shallow(
    <LanguagesFilter
      query={{ languages: null }}
      languages={languages}
      router={fakeRouter}
      facet={languagesFacet}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should render the languages facet with the selected languages', () => {
  const wrapper = shallow(
    <LanguagesFilter
      query={{ languages: ['java', 'cs'] }}
      value={['java', 'cs']}
      languages={languages}
      router={fakeRouter}
      facet={languagesFacet}
      isFavorite={true}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('Filter').shallow()).toMatchSnapshot();
});

it('should render maximum 10 languages in the searchbox results', () => {
  const wrapper = shallow(
    <LanguagesFilter
      query={{ languages: ['java', 'g'] }}
      value={['java', 'g']}
      languages={{
        ...languages,
        c: { key: 'c', name: 'c' },
        d: { key: 'd', name: 'd' },
        e: { key: 'e', name: 'e' },
        f: { key: 'f', name: 'f' },
        g: { key: 'g', name: 'g' },
        h: { key: 'h', name: 'h' },
        i: { key: 'i', name: 'i' },
        k: { key: 'k', name: 'k' },
        l: { key: 'l', name: 'l' }
      }}
      router={fakeRouter}
      facet={{ ...languagesFacet, g: 1 }}
      isFavorite={true}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
