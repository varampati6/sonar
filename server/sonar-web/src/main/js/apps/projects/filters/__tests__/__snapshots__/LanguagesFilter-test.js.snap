// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render maximum 10 languages in the searchbox results 1`] = `
<Filter
  facet={
    Object {
      "cs": 4,
      "g": 1,
      "java": 39,
      "js": 1,
    }
  }
  footer={
    <SearchableFilterFooter
      isFavorite={true}
      options={
        Array [
          Object {
            "label": "Flex",
            "value": "flex",
          },
          Object {
            "label": "PHP",
            "value": "php",
          },
          Object {
            "label": "Python",
            "value": "py",
          },
          Object {
            "label": "c",
            "value": "c",
          },
          Object {
            "label": "d",
            "value": "d",
          },
          Object {
            "label": "e",
            "value": "e",
          },
          Object {
            "label": "f",
            "value": "f",
          },
          Object {
            "label": "h",
            "value": "h",
          },
          Object {
            "label": "i",
            "value": "i",
          },
          Object {
            "label": "k",
            "value": "k",
          },
        ]
      }
      property="languages"
      query={
        Object {
          "languages": Array [
            "java",
            "g",
          ],
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Languages"
    />
  }
  highlightUnder={1}
  isFavorite={true}
  options={
    Array [
      "java",
      "cs",
      "g",
      "js",
    ]
  }
  property="languages"
  query={
    Object {
      "languages": Array [
        "java",
        "g",
      ],
    }
  }
  renderOption={[Function]}
  value={
    Array [
      "java",
      "g",
    ]
  }
/>
`;

exports[`should render the languages facet with the selected languages 1`] = `
<Filter
  facet={
    Object {
      "cs": 4,
      "java": 39,
      "js": 1,
    }
  }
  footer={
    <SearchableFilterFooter
      isFavorite={true}
      options={
        Array [
          Object {
            "label": "Flex",
            "value": "flex",
          },
          Object {
            "label": "PHP",
            "value": "php",
          },
          Object {
            "label": "Python",
            "value": "py",
          },
        ]
      }
      property="languages"
      query={
        Object {
          "languages": Array [
            "java",
            "cs",
          ],
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Languages"
    />
  }
  highlightUnder={1}
  isFavorite={true}
  options={
    Array [
      "java",
      "cs",
      "js",
    ]
  }
  property="languages"
  query={
    Object {
      "languages": Array [
        "java",
        "cs",
      ],
    }
  }
  renderOption={[Function]}
  value={
    Array [
      "java",
      "cs",
    ]
  }
/>
`;

exports[`should render the languages facet with the selected languages 2`] = `
<div
  className="search-navigator-facet-box"
  data-key="languages"
>
  <FilterHeader
    name="Languages"
  />
  <div
    className="search-navigator-facet-list"
  >
    <Link
      className="facet search-navigator-facet projects-facet active"
      data-key="java"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "languages": "cs",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          option={
            Object {
              "key": "java",
              "name": "Java",
            }
          }
          optionKey="java"
        />
      </span>
      <span
        className="facet-stat"
      >
        39
      </span>
    </Link>
    <Link
      className="facet search-navigator-facet projects-facet active"
      data-key="cs"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "languages": "java",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          option={
            Object {
              "key": "cs",
              "name": "C#",
            }
          }
          optionKey="cs"
        />
      </span>
      <span
        className="facet-stat"
      >
        4
      </span>
    </Link>
    <Link
      className="facet search-navigator-facet projects-facet"
      data-key="js"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/projects/favorite",
          "query": Object {
            "languages": "java,cs,js",
          },
        }
      }
    >
      <span
        className="facet-name"
      >
        <SearchableFilterOption
          option={
            Object {
              "key": "js",
              "name": "JavaScript",
            }
          }
          optionKey="js"
        />
      </span>
      <span
        className="facet-stat"
      >
        1
      </span>
    </Link>
  </div>
  <SearchableFilterFooter
    isFavorite={true}
    options={
      Array [
        Object {
          "label": "Flex",
          "value": "flex",
        },
        Object {
          "label": "PHP",
          "value": "php",
        },
        Object {
          "label": "Python",
          "value": "py",
        },
      ]
    }
    property="languages"
    query={
      Object {
        "languages": Array [
          "java",
          "cs",
        ],
      }
    }
    router={
      Object {
        "push": [Function],
      }
    }
  />
</div>
`;

exports[`should render the languages without the ones in the facet 1`] = `
<Filter
  facet={
    Object {
      "cs": 4,
      "java": 39,
      "js": 1,
    }
  }
  footer={
    <SearchableFilterFooter
      options={
        Array [
          Object {
            "label": "Flex",
            "value": "flex",
          },
          Object {
            "label": "PHP",
            "value": "php",
          },
          Object {
            "label": "Python",
            "value": "py",
          },
        ]
      }
      property="languages"
      query={
        Object {
          "languages": null,
        }
      }
      router={
        Object {
          "push": [Function],
        }
      }
    />
  }
  getFacetValueForOption={[Function]}
  halfWidth={false}
  header={
    <FilterHeader
      name="Languages"
    />
  }
  highlightUnder={1}
  options={
    Array [
      "java",
      "cs",
      "js",
    ]
  }
  property="languages"
  query={
    Object {
      "languages": null,
    }
  }
  renderOption={[Function]}
/>
`;
